### PR TITLE
CAPX-17 - New Mapping Button

### PR DESCRIPTION
**Pull request by sherakama**
# Ready for review
- Added create new mapping button and help text. 
- Broke out pages into their own file.
- Added button class to links so that they look like a button.

### CODE REVIEW

**Comment by sherakama** [comment]
@mpriest @pingers 

Please review when you have a moment. 

To test:
1. Navigate to the mapping landing page. See that create new mapping button looks like a button and that there is help text above it.

One significant change was moving the page callbacks to a pages.inc file.

**Comment by mpriest** [comment]
Fatal... stanford_capx_admin_config() is defined twice - Doh!

```
stanford_capx.module
349:function stanford_capx_admin_config() {

stanford_capx.pages.inc
16:function stanford_capx_admin_config() {
```

File: stanford_capx.module [hunk 1]
```diff
@@ -331,9 +334,10 @@ function stanford_capx_cron_orphan_queue_callback($item) {
 
 }
 
+// END CRON AND BATCH
+// /////////////////////////////////////////////////////////////////////////////
 
-
-// Pages
+// AUTOCOMPELTE
```
Comment (mpriest) [inline stanford_capx.module]: AUTOCOMPLETE

File: stanford_capx.pages.inc [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+ * @return string
+ *   HTML representation of this page.
+ */
+function stanford_capx_admin_config_connect() {
+
+  // Connection Status Block.
+  $connection_block = module_invoke('stanford_capx', 'block_view', 'connection_status');
+
+  // Content Output.
+  $output = "";
```
Comment (mpriest) [inline stanford_capx.pages.inc]: Not needed - just assign.

File: stanford_capx.pages.inc [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+  return $output;
+}
+
+/**
+ * Callback url and utility function for running the organizations sync.
+ * @return mixed
+ *   redirects back to page user was on.
+ */
+function stanford_capx_organizations_sync() {
+  drupal_set_message("Organization codes have been updated.", 'status');
```
Comment (mpriest) [inline stanford_capx.pages.inc]: Message and log are past tense before the sync runs. Maybe after they have successfully sync'd is a better time to say it worked?

File: stanford_capx.pages.inc [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+}
+
+/**
+ * Stanford Capx help page menu callback.
+ * @return string
+ *   HTML representation of this page.
+ */
+function stanford_capx_admin_config_help() {
+  $output = "";
+  $output .= "<p>Please create the functionality for this page</p>";
```
Comment (mpriest) [inline stanford_capx.pages.inc:100]: Translate it please.

File: stanford_capx.pages.inc [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+  return $output;
+}
+
+/**
+ * Stanford Capx mapper page menu callback.
+ * @return string
+ *   HTML representation of this page.
+ */
+function stanford_capx_admin_config_mapper() {
+  $output = "<p>After you have connected to CAP, create a Mapping to link CAP&#8217;s fields with your fields.</p>";
```
Comment (mpriest) [inline stanford_capx.pages.inc]: Should also translate these.

File: stanford_capx.pages.inc [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+  $output .= "<p><b>" . l(t('Create new mapper'), 'admin/config/capx/mapper/new', array('attributes' => array('class' => 'new-button button btn'))) . "</b></p>";
+  $output .= views_embed_view('capx_mappers');
+  return $output;
+}
+
+/**
+ * Stanford Capx create new mapper page menu callack.
```
Comment (mpriest) [inline stanford_capx.pages.inc]: "callback"

File: stanford_capx.pages.inc [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+  // Data browser block.
+  $block = block_load('stanford_capx', 'data_browser_launch');
+  $block_render = _block_get_renderable_array(_block_render_blocks(array($block)));
+  $output .= render($block_render);
+
+  // Get and render the mapper form.
+  $form = drupal_get_form('stanford_capx_mapper_form');
+  $output .= drupal_render($form);
+
+  return $output;
```
Comment (mpriest) [inline stanford_capx.pages.inc:137]: You really can just return the renderable array at this point and let the Drupal render it.
Just add the block and form to a renderable array and return it.

File: stanford_capx.pages.inc [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+/**
+ * Stanford Capx create new mapper page menu callack.
+ * @return string
+ *   HTML representation of this page.
+ */
+function stanford_capx_admin_config_mapper_new() {
+  $output = "";
+
+  // Add some css.
+  drupal_add_css(drupal_get_path('module', 'stanford_capx') . "/css/stanford_capx.admin.css");
```
Comment (mpriest) [inline stanford_capx.pages.inc]: If you use the render array, you can just use #attached.
https://api.drupal.org/api/drupal/developer%21topics%21forms_api_reference.html/7#attached

File: stanford_capx.pages.inc [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+
+  return $output;
+}
+
+/**
+ * Stanford Capx edit mapper page menu callback.
+ * @return string
+ *   HTML representation of this page.
+ */
+function stanford_capx_admin_config_mapper_edit($mapper_machine_name) {
```
Comment (mpriest) [inline stanford_capx.pages.inc:145]: Same comments from "new" apply to "edit".

File: stanford_capx.pages.inc [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+  return $output;
+}
+
+/**
+ * Create new importer page menu callback.
+ * @return string
+ *   HTML representation of this page.
+ */
+function stanford_capx_admin_config_import_new() {
+  $output = "";
```
Comment (mpriest) [inline stanford_capx.pages.inc]: Can be simplified to:

``` php
return drupal_get_form('stanford_capx_importer_form');
```

File: stanford_capx.pages.inc [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+}
+
+/**
+ * Edit importer configuration form page menu callback.
+ * @return string
+ *   HTML representation of this page.
+ */
+function stanford_capx_admin_config_import_edit($importer_machine_name) {
+  $output = "";
+  global $_capx_importer;
```
Comment (mpriest) [inline stanford_capx.pages.inc]: Can't we just inject the importer as an argument to the form and not use globals? Applies to all places globals have been used I think.

File: stanford_capx.pages.inc [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+  batch_process(drupal_get_destination());
+
+}
+
+/**
+ * [stanford_capx_admin_config description]
+ * @return [type] [description]
+ */
+
+function stanford_capx_debug() {
```
Comment (mpriest) [inline stanford_capx.pages.inc:270]: Don't think we need this long term :)

File: stanford_capx.pages.inc (hunk 3 above)
Comment (sherakama) [inline stanford_capx.pages.inc]: This is a cheat to have the message displayed after the batch runs. I suppose I should move this to the batch complete callback.

File: stanford_capx.pages.inc (hunk 4 above)
Comment (sherakama) [inline stanford_capx.pages.inc:100]: When there is real content. For now it is a reminder to us.

File: stanford_capx.pages.inc (hunk 11 above)
Comment (sherakama) [inline stanford_capx.pages.inc]: Ya, I think that is a good idea.

File: stanford_capx.pages.inc (hunk 12 above)
Comment (sherakama) [inline stanford_capx.pages.inc:270]: Not ready to kill it just yet.

**Comment by sherakama** [comment]
@mpriest @pingers,

Latest round of updates for this posted. Please pull and have another review. 

Thank you!

File: includes/CAPx/Drupal/Organizations/Orgs.php [hunk 13]
```diff
@@ -114,6 +115,20 @@ public static function syncOrganizations($batch = FALSE) {
   }
 
   /**
+   * Callback for when organization batch has finished.
+   * @param  [type] $context [description]
+   * @return [type]          [description]
+   */
+  public static function syncOrganizationsBatchFinished($success, $results, $operations) {
+    if ($success) {
+      drupal_set_message("Organization codes have been updated.", 'status');
```
Comment (RoSk0) [inline includes/CAPx/Drupal/Organizations/Orgs.php]: Must be translated(use t()).

File: includes/CAPx/Drupal/Organizations/Orgs.php [hunk 14]
```diff
@@ -114,6 +115,20 @@ public static function syncOrganizations($batch = FALSE) {
...
+    }
+    else {
+      drupal_set_message("Organization codes update failed! Try again.", 'error');
```
Comment (RoSk0) [inline includes/CAPx/Drupal/Organizations/Orgs.php]: Must be translated(use t()).

File: stanford_capx.forms.inc [hunk 15]
```diff
@@ -185,12 +185,12 @@ function stanford_capx_forms_connect_form_submit($form, &$form_state) {
  * @param  [type] $form_state [description]
  * @return [type]             [description]
  */
-function stanford_capx_mapper_form($form, &$form_state) {
+function stanford_capx_mapper_form($form, &$form_state, $mapper = null) {
```
Comment (RoSk0) [inline stanford_capx.forms.inc]: JFYI: Drupal coding standard prefer NULL, FALSE and TRUE.

File: stanford_capx.info [hunk 16]
```diff
@@ -2,7 +2,7 @@ name = Stanford CAPx
 description = Provides the ability to import profiles into existing entity types
 core = 7.x
 package = Stanford CAPx
-project stanford_capx
+project = stanford_capx
```
Comment (RoSk0) [inline stanford_capx.info:5]: Take a look at https://www.drupal.org/node/542202#project for details about using "project" property.

File: stanford_capx.pages.inc [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+/**
+ * Settings page menu callback.
+ * @return string
+ *   HTML representation of this page.
+ */
+function stanford_capx_admin_config_settings() {
+
+  $has_orgs = \CAPx\Drupal\Organizations\Orgs::checkOrgs();
+  if (!$has_orgs) {
+    drupal_set_message('The organization codes are not available. You should sync with the API now.', 'warning');
```
Comment (RoSk0) [inline stanford_capx.pages.inc]: Must be translated(use t()).

File: stanford_capx.pages.inc [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+  $output .= views_embed_view('capx_mappers');
+  return $output;
+}
+
+/**
+ * Stanford Capx create new mapper page menu callback.
+ * @return string
+ *   HTML representation of this page.
+ */
+function stanford_capx_admin_config_mapper_new() {
```
Comment (RoSk0) [inline stanford_capx.pages.inc:121]: In general it's recommended to have one form for create/update operation for the entity and I don't see any issues here to two same forms. Even so stanford_capx_admin_config_mapper_new() and stanford_capx_admin_config_mapper_edit() is just a callbacks, we can use only one for both cases.

File: stanford_capx.pages.inc [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+
+  return $output;
+}
+
+/**
+ * Create new importer page menu callback.
+ * @return string
+ *   HTML representation of this page.
+ */
+function stanford_capx_admin_config_import_new() {
```
Comment (RoSk0) [inline stanford_capx.pages.inc:186]: There is no need to create custom page callback like this, just spefify 'drupal_get_form' as page callback with "array('stanford_capx_importer_form')" as page arguments in hook_menu().

File: stanford_capx.pages.inc [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+ * Stanford Capx edit mapper page menu callback.
+ * @return string
+ *   HTML representation of this page.
+ */
+function stanford_capx_admin_config_mapper_edit($mapper_machine_name) {
+  $output = array();
+  $mapper = capx_cfe_load_by_machine_name($mapper_machine_name, 'mapper');
+
+  if (!$mapper) {
+    throw new Exception("Could not load a mapper with provided machine name. Please check url.");
```
Comment (RoSk0) [inline stanford_capx.pages.inc]: Must be translated(use t()).

File: stanford_capx.pages.inc [hunk 21]
```diff
@@ -0,0 +1,284 @@
...
+ * Edit importer configuration form page menu callback.
+ * @return string
+ *   HTML representation of this page.
+ */
+function stanford_capx_admin_config_import_edit($importer_machine_name) {
+  $output = "";
+  $importer = CAPxImporter::loadImporter($importer_machine_name);
+
+  if (!$importer) {
+    throw new Exception("Could not load an importer with provided machine name. Please check url.");
```
Comment (RoSk0) [inline stanford_capx.pages.inc]: Must be translated(use t()).

File: stanford_capx.pages.inc [hunk 22]
```diff
@@ -0,0 +1,284 @@
...
+
+  // Sanity...
+  $machine_name = check_plain($machine_name);
+
+  // Find by the name that was passed in.
+  $importer = CAPxImporter::loadEntityImporter($machine_name);
+
+  // Nothing to see here.
+  if (!$importer) {
+    throw new Exception("Could not find an importer by that name.");
```
Comment (RoSk0) [inline stanford_capx.pages.inc]: Must be translated(use t()).

File: stanford_capx.pages.inc [hunk 23]
```diff
@@ -0,0 +1,284 @@
...
+
+  // Sanity...
+  $machine_name = check_plain($machine_name);
+
+  // Get the orphanator object.
+  $orphanator = CAPxImporter::getEntityOrphanator($machine_name);
+
+  // Nothing to see must throw up.
+  if (!$orphanator) {
+    throw new Exception("Could not find an entity importer orphanator by that name.");
```
Comment (RoSk0) [inline stanford_capx.pages.inc]: Must be translated(use t()).

File: stanford_capx.info (hunk 16 above)
Comment (sherakama) [inline stanford_capx.info:5]: Thanks for this heads up but we have a special case to use this. 

We use:
https://github.com/SU-SWS/erus | https://www.drupal.org/project/erus

This module hooks into our custom modules and ties them into the update status notifications using github's api and release system.

File: stanford_capx.pages.inc (hunk 18 above)
Comment (sherakama) [inline stanford_capx.pages.inc:121]: The biggest difference between the two callbacks is: 

if (!$mapper) {
    throw new Exception("Could not load a mapper with provided machine name. Please check url.");
  }

I agree there is some duplicate code between the two callbacks that could be simplified.

There is some upcoming work that will add some additional blocks of content to the edit page only. I find it easier to add this content through the callback function rather that tying it into the form and adding conditionals.

File: stanford_capx.pages.inc (hunk 19 above)
Comment (sherakama) [inline stanford_capx.pages.inc:186]: Understood,

Some additional work will be added to the edit form that I want outside of the form and controlled through the callback.

File: stanford_capx.info (hunk 16 above)
Comment (RoSk0) [inline stanford_capx.info:5]: Great, then we just need to add "project status url" property.

**Pull request merged**